In [40]:
import pandas as pd

# Introduction

The data can be found under the following link: https://www.kaggle.com/datasets/nelgiriyewithana/australian-vehicle-prices

The goal of the project is price prediction based on the input features

In [41]:
df = pd.read_csv('./data/Australian Vehicle Prices.csv')

In [42]:
df.head()

Brand    Year   Model               Car/Suv  \
0      Ssangyong  2022.0  Rexton  Sutherland Isuzu Ute   
1             MG  2022.0     MG3             Hatchback   
2            BMW  2022.0    430I                 Coupe   
3  Mercedes-Benz  2011.0    E500                 Coupe   
4        Renault  2022.0  Arkana                   SUV   

                                       Title UsedOrNew Transmission  \
0       2022 Ssangyong Rexton Ultimate (awd)      DEMO    Automatic   
1  2022 MG MG3 Auto Excite (with Navigation)      USED    Automatic   
2                      2022 BMW 430I M Sport      USED    Automatic   
3           2011 Mercedes-Benz E500 Elegance      USED    Automatic   
4                 2022 Renault Arkana Intens      USED    Automatic   

         Engine DriveType  FuelType FuelConsumption Kilometres   ColourExtInt  \
0  4 cyl, 2.2 L       AWD    Diesel  8.7 L / 100 km       5595  White / Black   
1  4 cyl, 1.5 L     Front   Premium  6.7 L / 100 km         16  Black / Black   
2    4 cyl, 2 L      Rear   Premium  6.6 L / 100 km       8472   Grey / White   
3  8 cyl, 5.5 L      Rear   Premium   11 L / 100 km     136517  White / Brown   
4  4 cyl, 1.3 L     Front  Unleaded    6 L / 100 km       1035   Grey / Black   

            Location CylindersinEngine   BodyType     Doors     Seats   Price  
0     Caringbah, NSW             4 cyl        SUV   4 Doors   7 Seats   51990  
1     Brookvale, NSW             4 cyl  Hatchback   5 Doors   5 Seats   19990  
2      Sylvania, NSW             4 cyl      Coupe   2 Doors   4 Seats  108988  
3  Mount Druitt, NSW             8 cyl      Coupe   2 Doors   4 Seats   32990  
4   Castle Hill, NSW             4 cyl        SUV   4 Doors   5 Seats   34990

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16734 entries, 0 to 16733
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Brand              16733 non-null  object 
 1   Year               16733 non-null  float64
 2   Model              16733 non-null  object 
 3   Car/Suv            16706 non-null  object 
 4   Title              16733 non-null  object 
 5   UsedOrNew          16733 non-null  object 
 6   Transmission       16733 non-null  object 
 7   Engine             16733 non-null  object 
 8   DriveType          16733 non-null  object 
 9   FuelType           16733 non-null  object 
 10  FuelConsumption    16733 non-null  object 
 11  Kilometres         16733 non-null  object 
 12  ColourExtInt       16733 non-null  object 
 13  Location           16284 non-null  object 
 14  CylindersinEngine  16733 non-null  object 
 15  BodyType           16452 non-null  object 
 16  Doors              151

# Data understanding

One of the biggest challange will be, that most of the feature vectors contains string values

In [44]:
print(f"Shape: {df.shape}\n\nColumns: {list(df.columns)}")

Shape: (16734, 19)

Columns: ['Brand', 'Year', 'Model', 'Car/Suv', 'Title', 'UsedOrNew', 'Transmission', 'Engine', 'DriveType', 'FuelType', 'FuelConsumption', 'Kilometres', 'ColourExtInt', 'Location', 'CylindersinEngine', 'BodyType', 'Doors', 'Seats', 'Price']


In [45]:
# checking the number of unique values in each column
df[df.columns].nunique()

Brand                   76
Year                    45
Model                  781
Car/Suv                618
Title                 8804
UsedOrNew                3
Transmission             3
Engine                 106
DriveType                5
FuelType                 9
FuelConsumption        157
Kilometres           14262
ColourExtInt           834
Location               618
CylindersinEngine       11
BodyType                10
Doors                   13
Seats                   13
Price                 3794
dtype: int64

In [46]:
# Check the missing values
df.isnull().sum()

Brand                   1
Year                    1
Model                   1
Car/Suv                28
Title                   1
UsedOrNew               1
Transmission            1
Engine                  1
DriveType               1
FuelType                1
FuelConsumption         1
Kilometres              1
ColourExtInt            1
Location              450
CylindersinEngine       1
BodyType              282
Doors                1604
Seats                1705
Price                   3
dtype: int64

The data shows that in 450 cases out of 16734 the location is not given. This might be just due to poor data maintanance, however it should be considered, that for some cases it was a sensitive information and was not published intentionally.

In the second step, the target column, the price column was examined and the 3 rows without price values were eliminated from the dataset. One of those rows seemed to have just NaN values.

In [47]:
df[df.Price.isnull()]

Brand    Year       Model                Car/Suv  \
2391      NaN     NaN         NaN                    NaN   
10156   Isuzu  2016.0       D-MAX             Ute / Tray   
11039  Nissan  2018.0  Pathfinder  City Jeep - Pre Owned   

                                        Title UsedOrNew Transmission  \
2391                                      NaN       NaN          NaN   
10156     2016 Isuzu D-MAX LS-U HI-Ride (4X4)      USED    Automatic   
11039  2018 Nissan Pathfinder Ti X-tronic 2WD      USED    Automatic   

           Engine DriveType  FuelType FuelConsumption Kilometres  \
2391          NaN       NaN       NaN             NaN        NaN   
10156  4 cyl, 3 L       4WD    Diesel  8.1 L / 100 km     147648   
11039           -     Front  Unleaded               -      52414   

         ColourExtInt               Location CylindersinEngine    BodyType  \
2391              NaN                    NaN               NaN         NaN   
10156  Silver / Black  Hoppers Crossing, VIC             4 cyl  Ute / Tray   
11039   White / Black         Southbank, VIC                 -       Wagon   

          Doors     Seats Price  
2391        NaN       NaN   NaN  
10156   2 Doors   4 Seats   NaN  
11039       NaN       NaN   NaN

In [48]:
# as the goal of the analysis is to predict the price of the car, we can drop the rows with missing values in the price column
df = df.dropna(subset=['Price'], axis=0)
# adding the missing values in the Location column with the string 'Unknown'
df['Location'].fillna(value='Unknown', inplace=True)

An additional preparation step is to extract the numeric informations from the string columns

In [49]:
df['Doors'] = df['Doors'].str.extract('(\d+)').astype(float)
df['Seats'] = df['Seats'].str.extract('(\d+)').astype(float)
df['FuelConsumption'] = df['FuelConsumption'].str.extract('([\d.]+) L / 100 km').astype(float)
df['CylindersinEngine'] = df['CylindersinEngine'].str.extract('(\d+)').astype(float)
df['Engine'] = df['Engine'].str.extract('([\d.]+)').astype(float)
df['Transmission'] = df['Transmission'].replace('-', 'Other')
df['FuelType'] = df['FuelType'].replace('-', 'Other')

In [50]:
df.describe()

Year        Engine  FuelConsumption  CylindersinEngine  \
count  16731.000000  15058.000000     15035.000000       15058.000000   
mean    2016.229155      4.423629         7.663632           4.423629   
std        5.248000      1.120687         2.324684           1.120687   
min     1940.000000      0.000000         0.000000           0.000000   
25%     2013.000000      4.000000         6.500000           4.000000   
50%     2017.000000      4.000000         7.600000           4.000000   
75%     2020.000000      4.000000         8.800000           4.000000   
max     2023.000000     12.000000        27.000000          12.000000   

              Doors         Seats  
count  15129.000000  15028.000000  
mean       4.011633      5.101477  
std        0.713248      1.119179  
min        2.000000      2.000000  
25%        4.000000      5.000000  
50%        4.000000      5.000000  
75%        4.000000      5.000000  
max       12.000000     22.000000

In [51]:
df.head()

Brand    Year   Model               Car/Suv  \
0      Ssangyong  2022.0  Rexton  Sutherland Isuzu Ute   
1             MG  2022.0     MG3             Hatchback   
2            BMW  2022.0    430I                 Coupe   
3  Mercedes-Benz  2011.0    E500                 Coupe   
4        Renault  2022.0  Arkana                   SUV   

                                       Title UsedOrNew Transmission  Engine  \
0       2022 Ssangyong Rexton Ultimate (awd)      DEMO    Automatic     4.0   
1  2022 MG MG3 Auto Excite (with Navigation)      USED    Automatic     4.0   
2                      2022 BMW 430I M Sport      USED    Automatic     4.0   
3           2011 Mercedes-Benz E500 Elegance      USED    Automatic     8.0   
4                 2022 Renault Arkana Intens      USED    Automatic     4.0   

  DriveType  FuelType  FuelConsumption Kilometres   ColourExtInt  \
0       AWD    Diesel              8.7       5595  White / Black   
1     Front   Premium              6.7         16  Black / Black   
2      Rear   Premium              6.6       8472   Grey / White   
3      Rear   Premium             11.0     136517  White / Brown   
4     Front  Unleaded              6.0       1035   Grey / Black   

            Location  CylindersinEngine   BodyType  Doors  Seats   Price  
0     Caringbah, NSW                4.0        SUV    4.0    7.0   51990  
1     Brookvale, NSW                4.0  Hatchback    5.0    5.0   19990  
2      Sylvania, NSW                4.0      Coupe    2.0    4.0  108988  
3  Mount Druitt, NSW                8.0      Coupe    2.0    4.0   32990  
4   Castle Hill, NSW                4.0        SUV    4.0    5.0   34990

In [58]:
df[df.Engine != df.CylindersinEngine].shape

(1673, 19)

In [60]:
df[df.Engine != df.CylindersinEngine]

Brand    Year      Model             Car/Suv  \
5             Toyota  2004.0     Estima      ON FOUR WHEELS   
10            Toyota  2009.0      HiAce             Carbarn   
12             Honda  2015.0       City      USED Dealer ad   
13            Toyota  2017.0      HiAce      USED Dealer ad   
17        Mitsubishi  2013.0  Outlander  Exclusive Autohaus   
...              ...     ...        ...                 ...   
16712         Nissan  2010.0    Elgrand      USED Dealer ad   
16723         Suzuki  2022.0      Swift        NEW In stock   
16726         Toyota  2011.0     Estima          Topcar Rez   
16729         Toyota  2014.0    Alphard            SYC Cars   
16730  Mercedes-Benz  2012.0    S-Class      USED Dealer ad   

                                                   Title UsedOrNew  \
5                           2004 Toyota Estima T EDITION      USED   
10                          2009 Toyota HiAce (No Badge)      USED   
12     2015 Honda City GM VTi-L Sedan 4dr CVT 7sp 1.5...      USED   
13                                     2017 Toyota HiAce      USED   
17                          2013 Mitsubishi Outlander ES      USED   
...                                                  ...       ...   
16712                                2010 Nissan Elgrand      USED   
16723                          2022 Suzuki Swift GL Plus       NEW   
16726                                 2011 Toyota Estima      USED   
16729                                2014 Toyota Alphard      USED   
16730             2012 Mercedes-Benz S-Class CDI BlueTEC      USED   

      Transmission  Engine DriveType  FuelType  FuelConsumption Kilometres  \
5        Automatic     NaN     Other  Unleaded              NaN     160230   
10       Automatic     NaN     Other  Unleaded              NaN     148190   
12       Automatic     NaN     Other  Unleaded              NaN     181745   
13       Automatic     NaN     Other     Other              NaN     136296   
17          Manual     NaN     Other     Other              NaN     120020   
...            ...     ...       ...       ...              ...        ...   
16712    Automatic     NaN     Other     Other              NaN     132000   
16723    Automatic     NaN     Front  Unleaded              NaN         20   
16726    Automatic     NaN     Front  Unleaded              NaN      97000   
16729    Automatic     NaN     Front  Unleaded              NaN      66000   
16730    Automatic     NaN      Rear    Diesel              NaN     117432   

         ColourExtInt                Location  CylindersinEngine    BodyType  \
5            Grey / -        Saint Marys, NSW                NaN  Commercial   
10          White / -           Lidcombe, NSW                NaN  Commercial   
12          Black / -         Rouse Hill, NSW                NaN         NaN   
13      Silver / Grey           Lidcombe, NSW                NaN         NaN   
17        Blue / Grey          Blacktown, NSW                NaN       Wagon   
...               ...                     ...                ...         ...   
16712       White / -       Slacks Creek, QLD                NaN         NaN   
16723  Silver / Black      Indooroopilly, QLD                NaN   Hatchback   
16726      Silver / -  Holland Park West, QLD                NaN       Wagon   
16729        Grey / -            Rocklea, QLD                NaN  Commercial   
16730      Silver / -     Port Melbourne, VIC                NaN         NaN   

       Doors  Seats  Price  
5        NaN    NaN   9990  
10       NaN    NaN  42500  
12       NaN    NaN  11999  
13       NaN    NaN  38999  
17       NaN    NaN  13999  
...      ...    ...    ...  
16712    NaN    NaN  19000  
16723    NaN    NaN  28590  
16726    7.0    NaN  21989  
16729    7.0    NaN  29500  
16730    NaN    NaN  39888  

[1673 rows x 19 columns]

In [59]:
df.isnull().sum()

Brand                   0
Year                    0
Model                   0
Car/Suv                27
Title                   0
UsedOrNew               0
Transmission            0
Engine               1673
DriveType               0
FuelType                0
FuelConsumption      1696
Kilometres              0
ColourExtInt            0
Location                0
CylindersinEngine    1673
BodyType              281
Doors                1602
Seats                1703
Price                   0
dtype: int64

The Engine and the CylindersinEngine columns are after the transformation identical and both have the NaN values for the same entities. Therefore one could drop one of these columns. The more interesting information extraction would be to extract the volume parameters from the Engine column.